// @swagger-api/apidom-ns-json-schema-draft-7/src/refractor/visitors/json-schema/index.cjs
"use strict";

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault").default;
exports.__esModule = true;
exports.default = void 0;
var _tsMixer = require("ts-mixer");
var _ramda = require("ramda");
var _apidomNsJsonSchemaDraft = require("@swagger-api/apidom-ns-json-schema-draft-6");
var _JSONSchema = _interopRequireDefault(require("../../../elements/JSONSchema.cjs"));
/**
 * @public
 */

/**
 * @public
 */
class JSONSchemaVisitor extends (0, _tsMixer.Mixin)(_apidomNsJsonSchemaDraft.FixedFieldsVisitor, _apidomNsJsonSchemaDraft.ParentSchemaAwareVisitor, _apidomNsJsonSchemaDraft.FallbackVisitor) {
  constructor(options) {
    super(options);
    this.specPath = (0, _ramda.always)(['document', 'objects', 'JSONSchema']);
  }

  // eslint-disable-next-line class-methods-use-this
  get defaultDialectIdentifier() {
    return 'http://json-schema.org/draft-07/schema#';
  }
  ObjectElement(objectElement) {
    this.element = new _JSONSchema.default();
    this.handleDialectIdentifier(objectElement);
    this.handleSchemaIdentifier(objectElement);

    // for further processing consider this Schema Element as parent for all embedded Schema Elements
    this.parent = this.element;
    return _apidomNsJsonSchemaDraft.FixedFieldsVisitor.prototype.ObjectElement.call(this, objectElement);
  }
  BooleanElement(booleanElement) {
    const result = this.enter(booleanElement);
    this.element.classes.push('boolean-json-schema');
    return result;
  }
  handleDialectIdentifier(objectElement) {
    return _apidomNsJsonSchemaDraft.JSONSchemaVisitor.prototype.handleDialectIdentifier.call(this, objectElement);
  }
  handleSchemaIdentifier(objectElement, identifierKeyword = '$id') {
    return _apidomNsJsonSchemaDraft.JSONSchemaVisitor.prototype.handleSchemaIdentifier.call(this, objectElement, identifierKeyword);
  }
}
var _default = exports.default = JSONSchemaVisitor;
